packages/mask/src/plugins/Referral/SNSAdaptor/BuyToFarm.tsx

```diff
@@ -0,0 +1,228 @@
+import { makeStyles, useCustomSnackbar } from '@masknet/theme'
+import { useRemoteControlledDialog } from '@masknet/shared-base-ui'
+import { Typography, Box, Tab, Tabs, Grid, Divider } from '@mui/material'
+import { TabContext, TabPanel } from '@mui/lab'
+import { v4 as uuid } from 'uuid'
+
+import { useI18N } from '../../../utils'
+import { useRequiredChainId } from '../hooks/useRequiredChainId'
+import { PluginReferralMessages, SelectTokenUpdated } from '../messages'
+import { PluginTraderMessages } from '../../Trader/messages'
+import { farmsService, proofOfRecommendationService } from '../Worker/services'
+import { toChainAddressEthers, getFarmsRewardData } from '../helpers'
+import { MASK_REFERRER, SWAP_CHAIN_ID } from '../constants'
+import { TabsCreateFarm, TransactionStatus, PageInterface, PagesType, TabsReferralFarms } from '../types'
+import type { Coin } from '../../Trader/types'
+
+import ActionButton from '../../../extension/options-page/DashboardComponents/ActionButton'
+import { EthereumChainBoundary } from '../../../web3/UI/EthereumChainBoundary'
+import { MyFarms } from './MyFarms'
+import { TokenSelectField } from './shared-ui/TokenSelectField'
+import { RewardDataWidget } from './shared-ui/RewardDataWidget'
+import { SponsoredFarmIcon } from './shared-ui/icons/SponsoredFarm'
+
+import { useTabStyles, useSharedStyles } from './styles'
+
+const useStyles = makeStyles()((theme) => ({
+    container: {
+        flex: 1,
+        height: '100%',
+    },
+    tab: {
+        maxHeight: '100%',
+        height: '100%',
+        overflow: 'auto',
+        padding: `${theme.spacing(3)} 0`,
+    },
+    tabs: {
+        width: '288px',
+    },
+    subtitle: {
+        margin: '12px 0 24px',
+    },
+    typeNote: {
+        marginBottom: '24px',
+        '& b': {
+            margin: '0 4px 0 8px',
+            fontWeight: 600,
+        },
+    },
+}))
+
+export function BuyToFarm(props: PageInterface) {
+    const { t } = useI18N()
+    const { classes } = useStyles()
+    const { classes: tabClasses } = useTabStyles()
+    const { classes: sharedClasses } = useSharedStyles()
+    const currentChainId = useChainId()
+    const requiredChainId = useRequiredChainId(currentChainId)
+    const web3 = useWeb3()
+    const account = useAccount()
+    const { showSnackbar } = useCustomSnackbar()
+    const { ERC20 } = useTokenListConstants()
+
+    const [tab, setTab] = useState<string>(TabsCreateFarm.NEW)
+    const [id] = useState(uuid())
+    const [token, setToken] = useState<FungibleTokenDetailed>()
+    const { setDialog: setSelectTokenDialog } = useRemoteControlledDialog(
+        PluginReferralMessages.selectTokenUpdated,
+        useCallback(
+            (ev: SelectTokenUpdated) => {
+                if (ev.open || !ev.token || ev.uuid !== id) return
+                setToken(ev.token)
+            },
+            [id, setToken],
+        ),
+    )
```

Comment:
I have sent this patch in https://github.com/DimensionDev/Maskbook/pull/6153

Reply:
Our design requires more than just title and whitelist.

For now, we need to add an icon(farm type icon) to each token.
Shortly, we will need `Estimated APR` and filter by farm type in the future.
<img width="511" alt="Screenshot 2022-04-27 at 10 24 21" src="https://user-images.githubusercontent.com/32291396/165487060-396d23c5-f6d0-4aa2-b94b-888702402c88.png">

Reply:
@UncleBill Thanks, notify me, please.

Reply:
@UncleBill In order to avoid blockers at this stage, can we push our current solution and refactor later when this solution is available?

Reply:
> @UncleBill In order to avoid blockers at this stage, can we push our current solution and refactor later when this solution is available?

I think it's acceptable. Since I don't have to implement it nowadays.